arbnode/batch_poster.go

```diff
@@ -538,8 +530,7 @@ func AccessList(opts *AccessListOpts) types.AccessList {
 
 // Adds a block merkle proof to an Espresso justification, providing a proof that a set of transactions
 // hashes to some light client state root.
-func (b *BatchPoster) addEspressoBlockMerkleProof(
-	ctx context.Context,
+func (b *BatchPoster) checkEspressoValidation(
```

Comment:
I'm a bit worried about the espresso msg check here. With moving the escape hatch to the batch poster we also have no need to distinguish messages in the the sequencer anymore. This means we won't be prepending the espresso message byte tag to the l2 message before it reaches this method. This becomes complicated because if we add these bytes to the message here, we don't actually update the message in the transaction streamers DB, leading to calls to `arbos.ParseEspressoMessage()` to fail.

I think this points to a larger problem of the transaction streamer having overloaded responsibilities. That is, handle mutating it's messages in place in it's db so that we have a singular stream of messages that the batch poster reads from. 

Perhaps we migrate the work that the transaction streamer does for submitting to a separate thread in a component that only the batch poster holds. This thread would submit transactions to espresso, poll them for finality, and then serve espresso finalized transactions to the batch poster via a different DB. This avoids the weirdness with what component is holding the data.

This way in the batch posters main loop, we can process from the transaction streamers DB, and then send it down the espresso logic path with this new espresso transaction submitter, or the normal path. After this we could add it to a new queue for messages to be included in the batch, Once that queue is full, we then build and post the batch. 

Anyways, I suppose this is a problem to worry about in my own pr :sweat_smile:

Reply:
I am not sure I understand this, where are we adding bytes to messages?

Reply:
For escape hatch, can we not just skip `checkEspressoValidation(`  function if hotshot is not live?

Reply:
Thank you for your thought! It is helpful.
I do think the current flow in this PR is also strange (but better than previous) because the tx-streamer also has a thread to poll and do the validation. And as you said, the tx-streamer actually has too much responsibility, which also makes people confused.
The ideal flow for me is everything is done in batch poster and for the escape hatch, we can just use a simple condition.
I will continue refactoring something and feel free to work on this branch.